spec/unit/ssl/ssl_provider_spec.rb

```diff
@@ -0,0 +1,341 @@
+    File.read(my_fixture(name))
+  end
+
+  def cert(name)
+    OpenSSL::X509::Certificate.new(pem_content(name))
+  end
+
+  def crl(name)
+    OpenSSL::X509::CRL.new(pem_content(name))
+  end
+
+  def key(name)
+    OpenSSL::PKey::RSA.new(pem_content(name))
+  end
+
+  def request(name)
+    OpenSSL::X509::Request.new(pem_content(name))
+  end
+
+  let(:global_cacerts) { [ cert('ca.pem'), cert('intermediate.pem') ] }
+  let(:global_crls) { [ crl('crl.pem'), crl('intermediate-crl.pem') ] }
+  let(:wrong_key) { OpenSSL::PKey::RSA.new(512) }
+
+  context 'when creating an insecure context' do
+    let(:sslctx) { subject.create_insecure_context }
+
+    it 'has an empty list of trusted certs' do
+      expect(sslctx.trusted_certs).to eq([])
+    end
+
+    it 'has an empty list of crls' do
+      expect(sslctx.crls).to eq([])
+    end
+
+    it 'has an empty chain' do
+      expect(sslctx.client_chain).to eq([])
+    end
+
+    it 'has a nil private key and cert' do
+      expect(sslctx.private_key).to be_nil
+      expect(sslctx.client_cert).to be_nil
+    end
+
+    it 'does not authenticate the server' do
+      expect(subject.create_insecure_context.verify_peer).to eq(false)
+    end
+  end
+
+  context 'when creating an root ssl context with CA certs' do
+    let(:config) { { cacerts: [], crls: [], revocation: false } }
+
+    it 'accepts empty list of certs and crls' do
+      sslctx = subject.create_root_context(config)
+      expect(sslctx.trusted_certs).to eq([])
+      expect(sslctx.crls).to eq([])
+    end
+
+    it 'accepts valid root certs' do
+      certs = [cert('ca.pem')]
+      sslctx = subject.create_root_context(config.merge(cacerts: certs))
+      expect(sslctx.trusted_certs).to eq(certs)
+    end
+
+    it 'accepts valid intermediate certs' do
+      certs = [cert('ca.pem'), cert('intermediate.pem')]
+      sslctx = subject.create_root_context(config.merge(cacerts: certs))
+      expect(sslctx.trusted_certs).to eq(certs)
+    end
+
+    it 'accepts expired CA certs' do
+      expired = [cert('ca.pem'), cert('intermediate.pem')]
+      expired.each { |x509| x509.not_after = Time.at(0) }
+
+      sslctx = subject.create_root_context(config.merge(cacerts: expired))
+      expect(sslctx.trusted_certs).to eq(expired)
+    end
+  end
+
+  context 'when creating an ssl context with crls' do
+    let(:config) { { cacerts: global_cacerts, crls: global_crls} }
+
+    it 'accepts valid CRLs' do
+      certs = [cert('ca.pem')]
+      crls = [crl('crl.pem')]
+      sslctx = subject.create_root_context(config.merge(cacerts: certs, crls: crls))
+      expect(sslctx.crls).to eq(crls)
+    end
+
+    it 'accepts valid CRLs for intermediate certs' do
+      certs = [cert('ca.pem'), cert('intermediate.pem')]
+      crls = [crl('crl.pem'), crl('intermediate-crl.pem')]
+      sslctx = subject.create_root_context(config.merge(cacerts: certs, crls: crls))
+      expect(sslctx.crls).to eq(crls)
+    end
+
+    it 'accepts expired CRLs' do
```

Comment:
Is there anything it does not accept?

Reply:
The error cases are covered in the `create_context` section below. I could duplicate the tests (or use shared examples), but that feels like overkill.

Reply:
We looked at this more, and in this case, ruby will raise if `add_cert` is called with something other than a cert, and similar for `add_crl`.